spec/index.bs

```diff
@@ -5521,6 +5521,67 @@ dictionary GPUCommandBufferDescriptor : GPUObjectDescriptorBase {
 
 # Command Encoding # {#command-encoding}
 
+## <dfn interface>GPUCommandsMixin</dfn> ## {#commands-mixin}
+
+{{GPUCommandsMixin}} defines state common to all interfaces which encode commands.
+It has no methods.
+
+<script type=idl>
+interface mixin GPUCommandsMixin {
+};
+</script>
+
+{{GPUCommandsMixin}} adds the following internal slots to interfaces which include it:
+
+<dl dfn-type=attribute dfn-for=GPUCommandsMixin>
+    : <dfn>\[[state]]</dfn>, of type [=encoder state=]
+    ::
+        The current state of the {{GPUCommandEncoder}}, initially set to "[=encoder state/open=]".
+
+    : <dfn>\[[commands]]</dfn>, of type [=list=]&lt;[=GPU command=]&gt;
+    ::
+        A [=list=] of [=GPU commands=] to be executed on the [=Queue timeline=] when a
+        {{GPUCommandBuffer}} containing these commands is submitted.
+</dl>
+
+The <dfn dfn>encoder state</dfn> may be one of the following:
+
+<dl dfn-type=dfn dfn-for="encoder state">
+    : "<dfn>open</dfn>"
+    ::
+        The encoder is available to encode new commands.
+
+    : "<dfn>locked</dfn>"
+    ::
+        The encoder cannot be used, because it is locked by a child encoder: it is a
+        {{GPUCommandEncoder}}, and a {{GPURenderPassEncoder}} or {{GPUComputePassEncoder}} is active.
+        The encoder becomes "[=encoder state/open=]" again when the pass is ended.
+
+        Any command issued in this state makes the encoder [=invalid=].
+
+    : "<dfn>ended</dfn>"
+    ::
+        The encoder has been ended and new commands can no longer be encoded.
+
+        Any command issued in this state generates a {{GPUValidationError}}.
+</dl>
+
+<div algorithm>
+    To <dfn abstract-op>Validate the encoder state</dfn> of {{GPUCommandsMixin}} |encoder|:
+
+    If |encoder|.{{GPUCommandsMixin/[[state]]}} is:
+    <dl class=switch>
+        : "[=encoder state/open=]"
+        :: Return `true`.
+
+        : "[=encoder state/locked=]"
+        :: Make |encoder| [=invalid=], and return `false`.
```

Comment:
yikes, this looks suspicious. We have plenty of "Validate xxx" algorithms and all of them are "pure" as expected from the name. But this one is actually mutating the encoder (by making it invalid). Can we refactor this to avoid mutation?

Reply:
I'd prefer to keep the mutation here rather than duplicating it in EVERY command. But I can try to think of another name for the algorithm.

Reply:
Now that the algorithm isn't called "validate xxx" this is fine